### ISSUE TITLE
Correct way to shuffle, batch and shard WebDataset

### ISSUE THREAD

**Issue by austinmw** [pytorch/data#965]
### 📚 The doc issue

Hi, the [docs on the WebDataset decoder](https://pytorch.org/data/main/generated/torchdata.datapipes.iter.WebDataset.html) give the following example:

```python
>>> from torchdata.datapipes.iter import FileLister, FileOpener
>>>
>>> def decode(item):
>>>     key, value = item
>>>     if key.endswith(".txt"):
>>>         return key, value.read().decode("utf-8")
>>>     if key.endswith(".bin"):
>>>         return key, value.read().decode("utf-8")
>>>
>>> datapipe1 = FileLister("test/_fakedata", "wds*.tar")
>>> datapipe2 = FileOpener(datapipe1, mode="b")
>>> dataset = datapipe2.load_from_tar().map(decode).webdataset()
>>> for obj in dataset:
>>>     print(obj)
```

However this doesn't include demonstrating the proper location for shuffling, sharding and batching the dataset.

### Suggest a potential alternative/fix

Could you please let me know where to place `.shuffle`, `.batch` and `.sharding_filter` in this pipeline?

**Comment by ejguan** [comment]
Thanks for reporting the documentation issue.

The following should be the recommended way:
```py
>>> datapipe1 = FileLister("test/_fakedata", "wds*.tar")
>>> datapipe1 = datapipe1.shuffle().sharding_filter()
>>> datapipe2 = FileOpener(datapipe1, mode="b")
>>> dataset = datapipe2.load_from_tar().map(decode).webdataset().batch(batch_size)
>>> for obj in dataset:
>>>     print(obj)
```

**Comment by austinmw** [comment]
Thanks for your quick response! One more question if you don't mind, does `DataLoader2` and `PrototypeMultiProcessingReadingService` function with `webdataset`? I didn't notice any speed improvements between these two versions:

```python
%%time

for batch in datapipe:
    pass
```

versus

```python
%%time

reading_service = PrototypeMultiProcessingReadingService(num_workers=8)
dl = DataLoader2(datapipe=dataset, reading_service=reading_service)

for batch in dl:
    pass
````

**Comment by ejguan** [comment]
The speed improvement might not be significant if your pipeline is relatively simple, because Multiprocessing also has overhead.

**Comment by austinmw** [comment]
I see, thanks!
